lib/livebook_web/live/app_helpers.ex

```diff
@@ -79,4 +83,353 @@ defmodule LivebookWeb.AppHelpers do
+              |> JS.add_class("", transition: {"tooltip top", "", ""}, time: 2000)
+            }
+          >
+            <.remix_icon icon="clipboard-line" class="align-middle mr-1 text-xs" />
+            <span class="font-normal text-xs">Copy source</span>
+          </button>
+        </div>
+
+        <.code_preview source_id="dockerfile-source" source={@dockerfile} language="dockerfile" />
+      </div>
+
+      <div class="text-gray-700">
+        To test the deployment locally, go the the notebook directory, save the Dockerfile, then run:
+      </div>
+
+      <.code_preview
+        source_id="dockerfile-cmd"
+        source={
+          ~s'''
+          docker build -t my-app .
+          docker run --rm -p 8080:8080 -p 8081:8081 my-app
+          '''
+        }
+        language="text"
+      />
+
+      <p class="text-gray-700 py-2">
+        You may additionally perform the following optional steps:
+      </p>
+
+      <ul class="text-gray-700 space-y-3">
+        <li :if={Hubs.Provider.type(@hub) == "team"} class="flex gap-2">
+          <div><.remix_icon icon="arrow-right-line" class="text-gray-900" /></div>
+          <span>
+            you may remove the default value for <code>TEAMS_KEY</code>
+            from your Dockerfile and set it as a build argument in your deployment
+            platform
+          </span>
+        </li>
+        <li class="flex gap-2">
+          <div><.remix_icon icon="arrow-right-line" class="text-gray-900" /></div>
+          <span>
+            if you want to debug your deployed notebooks in production, you may
+            set the <code>LIVEBOOK_PASSWORD</code> environment variable with a
+            value of at least 12 characters of your choice
+          </span>
+        </li>
+      </ul>
+    </div>
+    """
+  end
+
+  defp zta_metadata(nil), do: nil
+
+  defp zta_metadata(zta_provider) do
+    Enum.find(Livebook.Config.identity_providers(), &(&1.type == zta_provider))
+  end
+
+  @doc """
+  Builds Dockerfile definition for app deployment.
+  """
+  @spec build_dockerfile(
+          map(),
+          Hubs.Provider.t(),
+          list(Livebook.Secrets.Secret.t()),
+          list(Livebook.FileSystem.t()),
+          Livebook.FileSystem.File.t() | nil,
+          list(Livebook.Notebook.file_entry()),
+          list(Livebook.Session.Data.secrets())
+        ) :: String.t()
+  def build_dockerfile(config, hub, hub_secrets, hub_file_systems, file, file_entries, secrets) do
+    base_image = Enum.find(Livebook.Config.docker_images(), &(&1.tag == config.docker_tag))
+
+    image = """
+    FROM ghcr.io/livebook-dev/livebook:#{base_image.tag}
+    """
+
+    image_envs = format_envs(base_image.env)
+
+    hub_config =
+      case Hubs.Provider.type(hub) do
+        "team" ->
+          format_team_hub_config(config, hub, hub_secrets, hub_file_systems)
+
+        "personal" ->
+          format_personal_hub_config(config, hub, hub_secrets, secrets)
+      end
```

Comment:
My suggestion is to have something like this:

```suggestion
      format_hub(Hubs.Provider.type(hub), config, hub, hub_secrets, hub_file_systems, secrets)
```

And keep the decision of what happens where inside that function. That's because if eventually we decide to extract that into the hub, the logic is encapsulated. It is fine to keep it as is though.

Another follow up question is: if we are deploying a single notebook, and we are using the teams hub, wouldn't it be better to only include the used secrets as well instead of all?

Reply:
I am fine either way as well. Your call. Whatever leads to fewer code I guess, always pruning for a single notebook should be fewer conditionals?